rpc, wallet: addhdseed, infer seed when importing descriptor with xpub

Sjors opened the pull request:
Builds on top of #25907 and one commit from #22341.

First this PR introduces a descriptor wallet RPC `addhdseed`. Similar to its legacy wallet counterpart `sethdseed` it either generates a fresh random seed, for use on a blank wallet, or the user can provide a WIF. Unlike `sethdseed` this RPC call does not add keys / fill the keypool.

This allows a user to create a blank wallet (`createwallet`), provide it with a seed (`addhdseed`) and then craft custom descriptors. (An alternative approach would to add a `nodescriptors` argument to `createwallet`, but this RPC seems useful anyway)

An example use case of this is setting up a multisig between Core and a hardware wallet. After adding the seed to an otherwise empty wallet, the user would call `getxpub m/87'/0'/0'`. They would then combine this `xpub` and origin info with the one from their hardware wallet (e.g. using `hwi --fingerprint .... getxpub m/87'/0'/0'`  and craft a descriptor like `wsh(sorted_multi(2, [...../87'/0'/0']our_xpub, [..../87'/0'/0']their_xpub))`, and import that. The resulting wallet will be able to sign its part of the transaction.

The second part of this PR enables to ability descriptors, such as in the multisig example above, in such a way that any `fingerprint` + `xpub` that is covered by our hd seed(s) can be signed for.

Sjors committed:
wallet: GetExtPubKey()

achow101 committed:
descriptor: Be able to get the pubkeys involved in a descriptor

achow101 committed:
wallet: Add KeyManager class

achow101 committed:
moveonly: move WalletStorage to separate file

achow101 committed:
descspkm: Add KeyManager to DescriptorScriptPubKeyMan and use for keys

achow101 committed:
descspkm: Encrypt with KeyManager instead of direct map access

achow101 committed:
descspkm: Track CKeyIDs of our keys

When DescriptorScriptPubKeyMan no longer manages its keys, it still
needs to know the IDs of its keys.

achow101 committed:
descspkm: Use KeyManager::LoadKey and LoadCryptedKey when loading

achow101 committed:
descspkm: Replace GetKeys with KeyManager::GetKeys

achow101 committed:
descspkm: Replace HavePrivateKeys with KeyManager::HavePrivateKeys()

achow101 committed:
keyman: Make some members private

achow101 committed:
wallet: Have KeyManager in CWallet rather than DescriptorScriptPubKeyMan

achow101 committed:
walletdb: Allow duplicate descriptor keys

If a descriptor (crypted) key is being written and one already exists,
make sure that the one being written and the one already on disk
match each other.

achow101 committed:
walletdb: Load keys into KeyManager directly

achow101 committed:
wallet: Add flag for using KeyManager

KeyManager will be a backwards compatible background upgrade to
descriptor wallets. A flag indicating that the upgrade has occurred is
added so that the upgrade (not yet implemented) will only happen once.

achow101 committed:
wallet: Use KeyManager to generate master key

achow101 committed:
walletdb: Implement upgrading a wallet to use KeyManager

achow101 committed:
descspkm: Remove unneeded key loading

Key management will be done entirely by KeyManager, so
DescriptorScriptPubKeyMan does not need key loading functions.

achow101 committed:
rpc: Add getxpub command

achow101 committed:
walletdb: Add HDKey records

achow101 committed:
walletdb: Add WriteKeyManKey and WriteCryptedKeyManKey

These functions write new key records for keys handled by a KeyManager

Rspigler commented:
Concept ACK! Thanks for continuing this work

Rspigler commented:
So cool! :rocket: 

I tried testing this as far as I could.

First Node:

Created an empty wallet.

`listdescriptors`:

> {
>   "wallet_name": "1",
>   "descriptors": [
>   ]
> }

:heavy_check_mark: 

`addhdseed`:

> {
>   "fingerprint": "2aba6847"
> }

`getxpub "m/87'/0'/0'"`:

> 
> {
>   "xpub": "xpub6Bp9vBsEbMipghoCm7fnWbSvXc9gVyZS5Duzfc8WgrEEbCC8mDeg8YGHAtihrLg1XdjJT1qKYgM4Q7REz8B7zgoH6mWEHdMPDwMCihsbSgK",
>   "fingerprint": "2aba6847",
>   "origin": "[2aba6847/87'/0'/0']"
> }

Now on to second node, since I don't have any HWW:

Created an empty wallet:

`listdescriptors`:

> {
>   "wallet_name": "2",
>   "descriptors": [
>   ]
> }

:heavy_check_mark: 

`addhdseed`:

> {
>   "fingerprint": "ad8a04e6"
> }

`getxpub "m/87'/0'/0'"`:

> {
>   "xpub": "xpub6DR65ocDPDwTphuxNogrNSWWsmUnFEpBWjhZABGxupZ65st8kaSWVmFripasF8in6JELXPNPbjwyqE8s7dqiGSYyrE9k5UGDF4XQCSALkqi",
>   "fingerprint": "ad8a04e6",
>   "origin": "[ad8a04e6/87'/0'/0']"
> }


Now back to first node, to combine into descriptor:

`wsh(sorted_multi(2,[2aba6847/87'/0'/0']xpub6Bp9vBsEbMipghoCm7fnWbSvXc9gVyZS5Duzfc8WgrEEbCC8mDeg8YGHAtihrLg1XdjJT1qKYgM4Q7REz8B7zgoH6mWEHdMPDwMCihsbSgK,[ad8a04e6/87'/0'/0']xpub6DR65ocDPDwTphuxNogrNSWWsmUnFEpBWjhZABGxupZ65st8kaSWVmFripasF8in6JELXPNPbjwyqE8s7dqiGSYyrE9k5UGDF4XQCSALkqi))`

I tried various formats/syntax for `importdescriptors`, but all failed.  I'm sure it is something I did on my end.  Can you help? For example:

`importdescriptors '[{ "desc": "wsh(sorted_multi(2,[2aba6847/87'/0'/0']xpub6Bp9vBsEbMipghoCm7fnWbSvXc9gVyZS5Duzfc8WgrEEbCC8mDeg8YGHAtihrLg1XdjJT1qKYgM4Q7REz8B7zgoH6mWEHdMPDwMCihsbSgK,[ad8a04e6/87'/0'/0']xpub6DR65ocDPDwTphuxNogrNSWWsmUnFEpBWjhZABGxupZ65st8kaSWVmFripasF8in6JELXPNPbjwyqE8s7dqiGSYyrE9k5UGDF4XQCSALkqi))>", "timestamp":now, "internal": true }`

Sjors commented:
@Rspigler in your examples the descriptors you importing are missing a derivation rule after the xpub, e.g. `xpub.../0/*` and they're missing a checksum (add a wrong one and it will tell you the right answers, e.g. `#00000000`)

Rspigler commented:
Those were silly mistakes, sorry.

Tried this:

`importdescriptors '[{ "desc": "wsh(sorted_multi(2,[3c9567fe/87'/0'/0'/0/*]xpub6DJT4N7LcSaLhKPSRkczg1nfD5z2wy1Kw48LAZHGTKDdWcjCc95wdRwQogYUX7M4fyzcXo3wQUrTH86RJ5w8cZ6T6nYdWe6guBN8SEBkPyD,[ad8a04e6/87'/0'/0'/0/*]xpub6DR65ocDPDwTphuxNogrNSWWsmUnFEpBWjhZABGxupZ65st8kaSWVmFripasF8in6JELXPNPbjwyqE8s7dqiGSYyrE9k5UGDF4XQCSALkqi))#00000000", "timestamp": now, "internal": "true" }]'`

But still received `Error: Error parsing JSON` error code.

(The first descriptor is different because I re-compiled on Node 1)

Sjors commented:
That's still the wrong format. The final `/0/*` needs to be _after_ the xpub. And "now" is missing quotes, whereas `true` should not be in quotes, which is probably what triggers the JSON parsing error. (I know it's tedious)

Rspigler commented:
Thanks for being patient with me.

Still getting errors but will continue to work on this myself rather than crowd the PR.

PastaPastaPasta commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1753,81 +1721,66 @@ void DescriptorScriptPubKeyMan::ReturnDestination(int64_t index, bool internal,
+    {
+        LOCK2(cs_desc_man, m_keyman.cs_keyman);
+        // Calculate the new range_end
+        int32_t new_range_end = std::max(m_wallet_descriptor.next_index + (int32_t)target_size, m_wallet_descriptor.range_end);
```
please convert this to a functional cast

achow101 committed:
wallet: Refactor function for single DescSPKM setup

We will need access to a function that sets up a singular
DescriptorSPKM, so refactor this out of the multiple DescriptorSPKM
setup function.

achow101 committed:
wallet, rpc: Add createwalletdescriptor RPC

achow101 committed:
wallet, descspkm: Move GetID into WalletDescriptor

achow101 committed:
wallet, descspkm: Refactor wallet descriptor generation to static func

achow101 committed:
walletdb: Allow overwriting keyman active hd key

achow101 committed:
wallet: Test for descriptor wallet sethdseed and createwalletdescriptor

Sjors commented:
There's some overlap between `addhdseed` that's introduced here and #24193, so I'll wait for that to evolve a bit and then maybe rebase on it.

achow101 committed:
wallet: Re-enable sethdseed for descriptors wallets

Descriptor wallets store an HD master key that is used for new
automatically generated descriptors. sethdseed is an existing RPC that
can be repurposed to allow the users to set that HD key whenever they
want.

Also fixes the whitespace of sethdseed. Best to review this with
--ignore-all-space

Sjors commented:
I'll rebase this on #25907 at some point.

Sjors commented:
Rebased on top of #25907. This uses a hardcoded xpriv, so don't bother testing just yet.

The approach now is that when calling `importdescriptors` is that if it finds missing private keys, it will try to derive them from the seed (using the new `keyman.GetActiveHDKey()` method). Will implement this derivation later using some stuff from #22341.

I'm not a huge fan of using string manipulation in the RPC code like this though.

Sjors commented:
It now works for the simplest case. Still need to clean up the part where I inject the master key xpub into the descriptor (hopefully that fixes the test too). I'd rather manipulate the `Descriptor` object directly than to manipulate strings and then re-parse.

Sjors committed:
wallet: match xpub in imported descriptor to existing seed

achow101 committed:
walletdb: Refactor deserialaization of keys with checksums

This will become shared later.

Sjors commented:
This is very out of date and it's not on my priority list at the moment. Happy to review if someone else takes it over.
